common/src/main/java/io/pravega/common/concurrent/Futures.java

```diff
@@ -599,6 +599,21 @@
         return result;
     }
 
+    /**
+     * Executes a loop using CompletableFutures, without invoking join()/get() on any of them or exclusively hogging a thread.
+     * This method also introduce a delay between each iteration of the loop.
+     *
+     * @param condition      A Supplier that indicates whether to proceed with the loop or not.
+     * @param loopBody       A Supplier that returns a CompletableFuture which represents the body of the loop. This
+     *                       supplier is invoked every time the loopBody needs to execute.
+     * @param delay           Delay in milliseconds.
+     * @param executor       An Executor that is used to execute the condition and the loop support code.
+     * @return A CompletableFuture that, when completed, indicates the loop terminated without any exception. If
+     * either the loopBody or condition throw/return Exceptions, these will be set as the result of this returned Future.
+     */
+    public static CompletableFuture<Void> loopWithDelay(Supplier<Boolean> condition, Supplier<CompletableFuture<Void>> loopBody, long delay, ScheduledExecutorService executor) {
```

Comment:
Is this really needed? This looks a lot like a one-off case which can be easily done by adding your `Futures.delayedFuture` as part of the `loopBody`.

I really prefer not having this extra method here, especially because you didn't add any unit tests for it.

Reply:
This PR is incomplete, i am yet to add all the unit tests.

Reply:
I think @andreipaduroiu is really making two comments. One is about unit tests that you responded to, @shiveshr . The other is that you implement the same functionality with existing calls. Could you address this latter point, please?

Reply:
I realized that i have seen that "composite" functionality quite useful and used often though in some tests.

Reply:
What I'm worried about is having this evolve into a method with so many overloads we won't be able to tell which one we really need to use. We begin by adding a "delay" for each iteration. Perhaps later we want an initial delay. Then we may want an iteration error handler. And so on. 

Making a parallel to the Java language, loops have a basic task: they loop. They do not do error handling or delay or  anything else. That was the goal of `Futures.loop`. I want a "primitive" that makes it easy for me to loop using CompletableFutures and not having to call `join` every time an iteration completes or write my own custom code.

I don't want to pollute this with custom features, such as iteration delays.

Reply:
Sure, moved it into a controller specific helper class.